archive/miu_language/basic.lean

```diff
@@ -0,0 +1,223 @@
+`M`, `I`, and `U`.
+
+Hofstadter's central question is: can the string `"MU"` be derived?
+
+It transpires that there is a simple decision procedure for this system. A string is derivable if
+and only if it starts with `M`, contains no other `M`s, and the number of `I`s in the string is
+congruent to 1 or 2 modulo 3.
+
+The principal aim of this project is to give a Lean proof that the derivability of a string is a
+decidable predicate.
+
+## The MIU System
+
+In Hofstadter's description, an _atom_ is any one of `M`, `I` or `U`. A _string_ is a finite
+sequence of zero or more symbols. To simplify notation, we write a sequence `[I,U,U,M]`,
+for example, as `IUUM`.
+
+The four rules of inference are:
+
+1. xI → xIU,
+2. Mx → Mxx,
+3. xIIIy → xUy,
+4. xUUy → xy,
+
+where the notation α → β is to be interpreted as 'if α is derivable, then β is derivable'.
+
+Additionally, he has an axiom:
+
+* `MI` is derivable.
+
+In Lean, it is natural to treat the rules of inference and the axiom on an equal footing via an
+inductive data type `derivable` designed so that `derviable x` represents the notion that the string
+`x` can be derived from the axiom by the rules of inference. The axiom is represented as a
+nonrecursive constructor for `derivable`. This mirrors the translation of Peano's axiom '0 is a
+natural number' into the nonrecursive constructor `zero` of the inductive type `nat`.
+
+## References
+
+* [Jeremy Avigad, Leonardo de Moura and Soonho Kong, _Theorem Proving in Lean_][avigad_moura_kong-2017]
+* [Douglas R Hofstadter, _Gödel, Escher, Bach_][Hofstadter-1979]
+
+## Tags
+
+miu, derivable strings
+
+-/
+
+
+namespace miu
+
+/-!
+### Declarations and instance derivations for `miu_atom` and `miustr`
+-/
+
+/--
+The atoms of MIU can be represented as an enumerated type in Lean.
+-/
+@[derive decidable_eq]
+inductive miu_atom : Type
+| M : miu_atom
+| I : miu_atom
+| U : miu_atom
+
+/-!
+The annotation `@[derive decidable_eq]` above assigns the attribute `derive` to `miu_atom`, through
+which Lean automatically derives that `miu_atom` is an instance of `decidable_eq`. The use of
+`derive` is crucial in this project and will lead to the automatic derivation of decidability.
+-/
+
+open miu_atom
+
+/--
+We show that the type `miu_atom` is inhabited, giving `M` (for no particular reason) as the default
+element.
+-/
+instance miu_atom_inhabited : inhabited miu_atom :=
+inhabited.mk M
+
+/--
+`miu_atom.repr` is the 'natural' function from `miu_atom` to `string`.
+-/
+def miu_atom.repr : miu_atom → string
+| M := "M"
+| I := "I"
+| U := "U"
+
+
+/--
+Using `miu_atom.repr`, we prove that ``miu_atom` is an instance of `has_repr`.
+-/
+instance : has_repr miu_atom :=
+⟨λ u, u.repr⟩
+
+/--
+For simplicity, an `miustr` is just a list of elements of type `miu_atom`.
+-/
+@[derive has_append]
+def miustr := list miu_atom
+
+/--
+We want to use list membership.
+-/
+instance : has_mem miu_atom miustr :=
```

Comment:
This can be replaced by changing the `derive` on `miustr` to `@[derive [has_append, has_mem miu_atom]]`